server/broadcast.go

```diff
@@ -82,7 +88,7 @@ func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment) {
 	vProfile := cxn.profile
 
 	cxn.lock.RLock()
-	sess := cxn.sess
+	sess := cxn.sessManager.selectSession(node, cxn.pl)
```

Comment:
This call can probably be moved down to line 117. That allows us to avoid selecting a session until we've completed any uploads to our local OS.

Reply:
True! This prompted me to notice something else - even if we move to line 117, if `uri, err := ios.SaveData(name, seg.Data)` fails (line 134), the session wouldn't be added back to our list (`cxn.sessManager.completeSession(sess)`), to no fault of the O. I'll add `cxn.sessManager.completeSession(sess)` to line 140 right before the return statement.